pkg/controller/deployment_inplace.go

```diff
@@ -0,0 +1,584 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove labels from the machines related to the inplace update.
+	for _, machine := range machinesWithUpdateSuccessfulLabel {
+		labelsToRemove := []string{
+			v1alpha1.LabelKeyNodeUpdateResult,
+		}
+
+		patchBytes, err := labelsutil.RemoveLabels(labelsToRemove)
+		if err != nil {
+			return err
+		}
+
+		klog.V(3).Infof("removing label %v from machine %s", labelsToRemove, machine.Name)
+		if err := dc.machineControl.PatchMachine(ctx, machine.Namespace, machine.Name, patchBytes); err != nil {
+			klog.Errorf("error while removing label  %v : %v", labelsToRemove, err)
+			return err
+		}
+	}
+
+	// updates nodes associated with the machines to remove update-related labels and annotations, and uncordons them.
+	for _, machine := range newMachines {
+		nodeName, ok := machine.Labels[v1alpha1.NodeLabelKey]
+		if !ok {
+			return fmt.Errorf("node label not found for machine %s: %w", machine.Name, err)
+		}
+
+		node, err := dc.nodeLister.Get(nodeName)
+		if err != nil {
+			return fmt.Errorf("failed to get node %s: %w", nodeName, err)
+		}
+
+		cond := nodeops.GetCondition(node, v1alpha1.NodeInPlaceUpdate)
+		if isUpdateNotSuccessful(cond, node.Labels) {
+			continue
+		}
+
+		// remove labels related to the inplace update.
+		delete(node.Labels, v1alpha1.LabelKeyNodeCandidateForUpdate)
+		delete(node.Labels, v1alpha1.LabelKeyNodeSelectedForUpdate)
+		delete(node.Labels, v1alpha1.LabelKeyNodeUpdateResult)
+		// remove annotations related to the inplace update.
+		delete(node.Annotations, v1alpha1.AnnotationKeyMachineUpdateFailedReason)
+
+		// uncordon the node since the inplace update is successful.
+		node.Spec.Unschedulable = false
+
+		_, err = dc.targetCoreClient.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to remove inplace labels/annotations and uncordon node %s: %w", node.Name, err)
+		}
+	}
+
+	for _, machineSet := range oldMachineSets {
+		// scale down the old machine set to the number of machines which is having the labelselector of the machine set.
+		oldMachines, err := dc.machineLister.List(labels.SelectorFromSet(machineSet.Spec.Selector.MatchLabels))
+		if err != nil {
+			return fmt.Errorf("failed to list machines for machine set %s: %w", machineSet.Name, err)
+		}
+
+		if len(oldMachines) < int(machineSet.Spec.Replicas) {
+			_, _, err := dc.scaleMachineSetAndRecordEvent(ctx, machineSet, int32(len(oldMachines)), deployment) // #nosec G115 (CWE-190) -- value already validated
+			if err != nil {
+				return fmt.Errorf("failed to scale down machine set %s: %w", machineSet.Name, err)
+			}
+			klog.V(3).Infof("scaled down machine set %s to %d", machineSet.Name, len(oldMachines))
+		}
+	}
+
+	return nil
+}
+
+func (dc *controller) reconcileNewMachineSetInPlace(ctx context.Context, oldMachineSets []*v1alpha1.MachineSet, newMachineSet *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
```

Comment:
The name of the method suggests that you are going to reconcile only new machineSet. That is not true, as you are also transferring machines from old to new and therefore you are also reconciling old machine set as a consequence.